src/node.js

```diff
@@ -390,6 +391,78 @@
     }
   };
 
+  startup.processPromises = function() {
+    // this should be kept in sync with PromiseRejectEvent in v8.h
+    var kPromiseRejectWithNoHandler = 0;
+    var kPromiseHandlerAddedAfterReject = 1;
+
+    var timers = NativeModule.require('timers');
+    var unhandledPromises = new WeakMap();
+    // A hack to work around uniterable WeakMaps.
+    var unhandledPromiseKeys = [];
+    var fullTurnQueued = false;
+
+    function queueFullTurn() {
+      if (!fullTurnQueued) {
+        fullTurnQueued = true;
+        timers.setImmediate(function() {
+          process.nextTick(afterFullTurn);
```

Comment:
Two reasons:
1.  After `setImmediate` callback some things can happen in the CURRENT event loop turn: other `setImmediate` callbacks, `nextTick` callbacks, microtasks

2. If you call `setImmediate` from, say, `setTimeout` callback or io callback it will actually be called in the end of the current event loop turn (it might be not the last thing though). If you call it from `setImmediate` callback it will be called in the next event loop turn.